materialize-boilerplate/stream-encode/parquet.go

```diff
@@ -0,0 +1,701 @@
+				return fmt.Errorf("writing date column: %w", err)
+			}
+		case LogicalTypeTime:
+			if err := writeColumn(colIdx, e.buffer, cw.(*file.Int64ColumnChunkWriter), getTimeVal); err != nil {
+				return fmt.Errorf("writing time column: %w", err)
+			}
+		case LogicalTypeTimestamp:
+			if err := writeColumn(colIdx, e.buffer, cw.(*file.Int64ColumnChunkWriter), getTimestampVal); err != nil {
+				return fmt.Errorf("writing timestamp column: %w", err)
+			}
+		case LogicalTypeInterval:
+			if err := writeColumn(colIdx, e.buffer, cw.(*file.FixedLenByteArrayColumnChunkWriter), getIntervalVal); err != nil {
+				return fmt.Errorf("writing interval column: %w", err)
+			}
+		default:
+			panic(fmt.Sprintf("attempted to encode unknown type: %d", f.DataType))
+		}
+
+		if err := cw.Close(); err != nil {
+			return fmt.Errorf("closing column writer: %w", err)
+		}
+	}
+
+	if err := rgWriter.Close(); err != nil {
+		return fmt.Errorf("closing row group writer: %w", err)
+	}
+
+	e.scratch.sizeBytes += int(rgWriter.TotalBytesWritten())
+	e.scratch.columnChunkCount += len(e.schema)
+	e.buffer = e.buffer[:0]
+	e.bufferSizeBytes = 0
+
+	return nil
+}
+
+type parquetValue interface {
+	int64 | int32 | float64 | bool | parquet.FixedLenByteArray | parquet.ByteArray
+}
+
+// columnBatchReader is a generic wrapper for reading a batch of values having type T from a column.
+// This interface is satisfied by any of the typed column readers from the Apache parquet package.
+type columnBatchReader[T parquetValue] interface {
+	// ReadBatch reads batchSize values from the column. values must be large enough to hold the
+	// number of values that will be read. defLvls and repLvls will be populated if not nil; they
+	// are not inputs to the operation but their populated values are necessary for interpreting the
+	// data read into values. total is the number of rows that were read; valuesRead is the actual
+	// number of physical values that were read excluding nulls.
+	ReadBatch(batchSize int64, values []T, defLvls []int16, repLvls []int16) (total int64, valuesRead int, err error)
+}
+
+// columnBatchWriter is a generic wrapper for writing a batch of values having type T to a column.
+// This interface is satisfied by any of the typed column writers from the Apache parquet package.
+type columnBatchWriter[T parquetValue] interface {
+	// WriteBatch writes a batch of repetition levels, definition levels, and values to the column.
+	// We don't currently support nested data structures (typed arrays being the only ones we
+	// reasonably could), so repLvls is always nil. The number of values in defLvls must equal the
+	// number of conceptual rows that are being written. The vals slice may contain a number of
+	// values less than or equal to the number of rows, as null values are omitted. Since we don't
+	// currently support nested data structures, a repLvl of 0 means the row at that corresponding
```

Comment:
did you mean to say "a defLevel of 0 means the row at that corresponding..." instead of "repLevel"? I don't see repLevel being used here in the code

Reply:
Yep! Thanks.